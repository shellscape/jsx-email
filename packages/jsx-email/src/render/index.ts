export * from './render';
export * from './jsx-to-string';
export * from './process';
