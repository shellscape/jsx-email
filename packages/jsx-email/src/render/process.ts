// Note: For some reason CI has a hard time with this, even though the dependency is installed, and
// it builds locally fine
// @ts-ignore
import type { Element, Parents, Root } from 'hast';
import prettyHtml from 'pretty';

import type { ProcessOptions } from '../types';

export { prettyHtml };

export const jsxEmailTags = ['jsx-email-cond'];

export const processHtml = async ({ html, minify, pretty }: ProcessOptions) => {
  const { rehype } = await import('rehype');
  const { default: stringify } = await import('rehype-stringify');
  const { visit } = await import('unist-util-visit');
  const docType =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const settings = {
    emitParseErrors: true
    // fragment: true
  };
  const reJsxTags = new RegExp(`<[/]?(${jsxEmailTags.join('|')})>`, 'g');

  function rehypeMoveStyle() {
    return function (tree: Root) {
      const matches: Array<[Parents, Element]> = [];
      let head: Element | undefined;

      visit(tree, 'element', (node, _, parent) => {
        if (node.tagName === 'head') {
          head = node;
        }

        if (
          parent &&
          node.tagName === 'style' &&
          (parent.type !== 'element' || parent.tagName !== 'head')
        ) {
          matches.push([parent, node]);
        }
      });

      if (head) {
        let index = -1;

        // eslint-disable-next-line no-plusplus
        while (++index < matches.length) {
          const match = matches[index];
          const siblings = match[0].children;
          siblings.splice(siblings.indexOf(match[1]), 1);
          head.children.push(match[1]);
        }
      }
    };
  }

  let processor = rehype().data('settings', settings).use(rehypeMoveStyle);

  if (minify) {
    const { default: rehypeMinify } = await import('rehype-preset-minify');
    processor = processor.use(rehypeMinify);
  }

  const doc = await processor
    .use(stringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
      closeEmptyElements: true,
      collapseEmptyAttributes: true
    })
    .process(html);

  let result = docType + String(doc).replace('<!doctype html>', '').replace('<head></head>', '');

  result = result.replace(reJsxTags, '');

  if (pretty) result = prettyHtml(result);

  return result;
};
