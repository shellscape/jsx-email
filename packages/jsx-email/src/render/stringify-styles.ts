/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { CSSProperties } from 'react';

import { escapeString } from './escape-string';

const UPPERCASE = /([A-Z])/g;

const MS = /^ms-/;

const UNITLESS_PROPS = new Set([
  'animationIterationCount',
  'columns',
  'columnCount',
  'flex',
  'flexGrow',
  'flexShrink',
  'fontWeight',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnStart',
  'gridRow',
  'gridRowEnd',
  'gridRowStart',
  'lineHeight',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom'
]);

const CACHE: Record<string, string> = {};

function hyphenateChar(char: string) {
  return `-${char.toLowerCase()}`;
}

function hyphenateString(prop: string) {
  return prop.replace(UPPERCASE, hyphenateChar).replace(MS, '-ms-');
}

/**
 * Converts a React style object to a string.
 * This is based on Hyperon's
 * [`stringifyStyles`](https://github.com/i-like-robots/hyperons/blob/main/src/stringify-styles.js).
 * @param styles The styles to stringify.
 * @returns A string representation of the styles, suitable for use in a `style` attribute.
 */
export function stringifyStyles(styles: CSSProperties) {
  const parts = [];
  for (const key of Object.keys(styles) as ReadonlyArray<keyof typeof styles>) {
    const value = styles[key];
    if (value != null) {
      const unit = typeof value === 'number' && value !== 0 && !UNITLESS_PROPS.has(key) ? 'px' : '';
      const normalized = CACHE[key] || (CACHE[key] = hyphenateString(key));
      parts.push(`${normalized}:${typeof value === 'number' ? value : escapeString(value)}${unit}`);
    }
  }

  return parts.join(';');
}
