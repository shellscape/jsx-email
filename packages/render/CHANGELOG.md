# @jsx-email/render ChangeLog

## v1.0.0

_2023-09-27_

### Breaking Changes

- chore!: initial release (4106d7e)