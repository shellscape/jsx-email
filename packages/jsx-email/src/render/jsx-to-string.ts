/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
import hash from 'hash-it';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

const promiseMap = new Map();

const wrapPromise = <TPromise extends Promise<any>>(promise: TPromise) => {
  let status = 'pending';
  let result: Awaited<TPromise>;
  const suspender = promise.then(
    (r) => {
      status = 'success';
      result = r;
    },
    (e) => {
      status = 'error';
      result = e;
    }
  );
  return {
    // eslint-disable-next-line consistent-return
    read() {
      if (status === 'pending') {
        throw suspender;
      } else if (status === 'error') {
        throw result;
      } else if (status === 'success') {
        return result;
      }
    }
  };
};

export const useData = <TData>(props: any, cb: () => Promise<TData>): TData => {
  const key = hash(props);
  let dataPromise;
  if (promiseMap.has(key)) {
    dataPromise = promiseMap.get(key);
  } else {
    dataPromise = wrapPromise(cb());
    promiseMap.set(key, dataPromise);
  }
  return dataPromise.read();
};

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error('Unsupported JSX element:', element as any);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }

    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
