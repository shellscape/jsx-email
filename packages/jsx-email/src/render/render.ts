import { htmlToText } from 'html-to-text';

import type { PlainTextOptions, RenderOptions } from '../types';

import { jsxToString } from './jsx-to-string';
import { processHtml } from './process';

export const renderPlainText = async (
  component: React.ReactElement,
  options?: PlainTextOptions
) => {
  const result = await jsxToString(component);
  return htmlToText(result, {
    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-skip="true"]' },
      { options: { linkBrackets: false }, selector: 'a' }
    ],
    ...options
  });
};

export const render = async (component: React.ReactElement, options?: RenderOptions) => {
  const { minify = false, plainText, pretty = false } = options || {};

  if (plainText) return renderPlainText(component, typeof plainText === 'object' ? plainText : {});

  let html = await jsxToString(component);

  html = await processHtml({ html, minify, pretty });

  return html;
};
