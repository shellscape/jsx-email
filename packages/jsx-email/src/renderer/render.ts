import { htmlToText } from 'html-to-text';

import { defineConfig, loadConfig, mergeConfig, type JsxEmailConfig } from '../config.js';
import { callHook, callProcessHook } from '../plugins.js';
import type { PlainTextOptions, RenderOptions } from '../types.js';

import { jsxToString } from './jsx-to-string.js';
import { getMovePlugin } from './move-style.js';
import { unescapeForRawComponent } from './raw.js';

export const jsxEmailTags = ['jsx-email-cond'];

export const renderPlainText = async (
  component: React.ReactElement,
  options?: PlainTextOptions
) => {
  const { formatters, selectors } = options || {};

  const result = await jsxToString(component);
  return htmlToText(result, {
    formatters: {
      raw: (elem, _walk, builder) => {
        if (elem.children.length && elem.children[0].type === 'comment') {
          builder.addInline(unescapeForRawComponent(elem.children[0].data!.trim()));
        }
      },
      ...formatters
    },

    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-skip="true"]' },
      { options: { linkBrackets: false }, selector: 'a' },
      {
        format: 'raw',
        options: {},
        selector: 'jsx-email-raw'
      },
      ...(selectors || [])
    ],
    ...options
  });
};

export const render = async (component: React.ReactElement, options?: RenderOptions) => {
  let config = await loadConfig();
  if (config.render.plainText || options?.plainText)
    return renderPlainText(
      component,
      typeof options?.plainText === 'object' ? options.plainText : {}
    );

  const renderOptions = { render: options };

  if (options) {
    // Note: structuredClone chokes on symbols
    const { symbol: _, ...cloneTarget } = config as any;
    const merged = await mergeConfig(cloneTarget, renderOptions);

    config = await defineConfig(merged);
  }

  let html = await jsxToString(component);

  html = await callHook({ config, hookType: 'beforeRender', html });
  html = await processHtml(config, html);
  html = await callHook({ config, hookType: 'afterRender', html });
  return html;
};

const processHtml = async (config: JsxEmailConfig, html: string) => {
  const { rehype } = await import('rehype');
  const { default: stringify } = await import('rehype-stringify');
  const docType =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const movePlugin = await getMovePlugin();
  const settings = { emitParseErrors: true };
  const reJsxTags = new RegExp(`<[/]?(${jsxEmailTags.join('|')})>`, 'g');

  // @ts-ignore: This is perfectly valid, see here: https://www.npmjs.com/package/rehype#examples
  const processor = rehype().data('settings', settings);

  processor.use(movePlugin);
  await callProcessHook({ config, processor });

  const doc = await processor
    .use(stringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
      closeEmptyElements: true,
      collapseEmptyAttributes: true
    })
    .process(html);

  let result = docType + String(doc).replace('<!doctype html>', '').replace('<head></head>', '');

  result = result.replace(reJsxTags, '');
  result = result.replace(/<jsx-email-raw.*?><!--(.*?)--><\/jsx-email-raw>/g, (_, p1) =>
    unescapeForRawComponent(p1)
  );

  return result;
};
