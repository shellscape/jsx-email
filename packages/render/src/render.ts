import { convert } from 'html-to-text';
import pretty from 'pretty';

import { jsxToString } from './jsx-to-string';

export interface Options {
  plainText?: boolean;
  pretty?: boolean;
}

export const render = (component: React.ReactElement, options?: Options) => {
  if (options?.plainText) {
    return renderAsPlainText(component, options);
  }
  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const markup = jsxToString(component);
  const document = `${doctype}${markup}`;

  if (options && options.pretty) {
    return pretty(document);
  }

  return document;
};

export const renderAsync = async (component: React.ReactElement, options?: Options) =>
  render(component, options);

const renderAsPlainText = (component: React.ReactElement, _options?: Options) =>
  convert(jsxToString(component), {
    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-id="@jsx-email/preview"]' }
    ]
  });
