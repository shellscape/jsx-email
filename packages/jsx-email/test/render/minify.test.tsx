// @ts-ignore
import React from 'react';

import { render } from '../../src/renderer/render.js';
import { Button } from '../../src/index.js';

describe('<Button> component', async () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('minifies outlook xml properly', async () => {
    const html = await render(
      <Button width={200} height={40}>
        {'Test message'}
      </Button>,
      { minify: true }
    );
    expect(html).toMatchSnapshot();
  });
});
