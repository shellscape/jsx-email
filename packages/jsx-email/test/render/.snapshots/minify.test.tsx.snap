// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button> component > minifies outlook xml properly 1`] = `
"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><head><table border=\\"0\\"cellpadding=\\"0\\"cellspacing=\\"0\\"role=\\"presentation\\"style=\\"border-collapse:collapse\\"width=\\"100%\\"><tr><td align=\\"left\\"><span><!--[if mso]>
            <v:roundrect href=\\"undefined\\" xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:w=\\"urn:schemas-microsoft-com:office:word\\" style=\\"height:40px;v-text-anchor:middle;width:200px;\\" arcsize=\\"0%\\"  strokeweight=\\"1px\\" fill=\\"false\\">
            <w:anchorlock/>
            <center style=\\"font-size:16px;\\">
            Test message
            </center></v:roundrect>
            <![endif]--></span><a style=\\"-webkit-text-size-adjust:none;border-radius:0;display:inline-block;font-size:16px;line-height:38px;max-width:200px;text-align:center;text-decoration:none;width:100%;mso-hide:all\\">Test message</a></table>"
`;
